ivadomed/loader/utils.py

```diff
@@ -722,149 +722,221 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
+        self.df = self.df[(~self.df['path'].str.contains('derivatives')
+                           & self.df['suffix'].str.contains('|'.join(self.contrast_lst))
+                           & self.df['extension'].str.contains('|'.join(self.extensions)))
+                           | (self.df['path'].str.contains('derivatives')
+                           & self.df['filename'].str.contains('|'.join(self.target_suffix)))]
+
+        # Add participant_id column, and metadata from participants.tsv file if present
+        # Uses pybids function
+        self.df['participant_id'] = "sub-" + self.df['subject']
+        if layout.get_collections(level='dataset'):
+            df_participants = layout.get_collections(level='dataset', merge=True).to_df()
+            df_participants.drop(['suffix'], axis=1, inplace=True)
+            self.df = pd.merge(self.df, df_participants, on='subject', suffixes=("_x", None), how='left')
+
+        # Add sample_id column if sample column exists, and add metadata from samples.tsv file if present
+        # TODO: use pybids function after BEP microscopy is merged in BIDS
+        if 'sample' in self.df:
+            self.df['sample_id'] = "sample-" + self.df['sample']
+        fname_samples = os.path.join(self.bids_path, "samples.tsv")
+        if os.path.exists(fname_samples):
+            df_samples = pd.read_csv(fname_samples, sep='\t')
+            self.df = pd.merge(self.df, df_samples, on=['participant_id', 'sample_id'], suffixes=("_x", None),
+                               how='left')
+
+        # Add metadata from all _sessions.tsv files, if present
+        # Uses pybids function
+        if layout.get_collections(level='subject'):
+            df_sessions = layout.get_collections(level='subject', merge=True).to_df()
+            df_sessions.drop(['suffix'], axis=1, inplace=True)
+            self.df = pd.merge(self.df, df_sessions, on=['subject', 'session'], suffixes=("_x", None), how='left')
+
+        # Add metadata from all _scans.tsv files, if present
+        # TODO: use pybids function after BEP microscopy is merged in BIDS
+        # TODO: verify merge behavior with EEG and DWI scans files, tested with anat and microscopy only
+        df_scans = pd.DataFrame()
+        for root, dirs, files in os.walk(self.bids_path):
+            for file in files:
+                if file.endswith("scans.tsv"):
+                    df_temp = pd.read_csv(os.path.join(root, file), sep='\t')
+                    df_scans = pd.concat([df_scans, df_temp], ignore_index=True)
+        if not df_scans.empty:
+            df_scans['filename'] = df_scans['filename'].apply(os.path.basename)
+            self.df = pd.merge(self.df, df_scans, on=['filename'], suffixes=("_x", None), how='left')
+
+        # TODO: check if other files are needed for EEG and DWI
+
+        # If indexing of derivatives is true
+        # Get list of subject files with available derivatives
+        if self.derivatives:
```

Comment:
Suggestion, code nested within this condition could be placed in another function (perhaps something like `get_subject_file_list()`

Reply:
Thanks for your suggestions, I implemented it as `get_subjects_with_derivatives` and I create another method specific to adding tsv files metadata `add_tsv_metadata`. Let me know if that makes sense. Thanks!